pkg/pgmodel/querier/query_builder.go

```diff
@@ -242,7 +242,29 @@ func canAttemptPushdown(metadata *promqlMetadata) bool {
 	path := metadata.path // PromQL AST.
 	queryHints := metadata.queryHints
 	selectHints := metadata.selectHints
-	return extension.ExtensionIsInstalled && queryHints != nil && !hasSubquery(path) && selectHints != nil
+
+	// We can't push down without hints
+	if queryHints == nil || selectHints == nil {
+		return false
+	}
+
+	// We can't do pushdowns without the extension
+	if !extension.ExtensionIsInstalled {
+		return false
+	}
+
+	// We can't handle subqueries in pushdowns
+	if hasSubquery(path) {
+		return false
+	}
+
+	// We can't handle offsets in pushdowns
+	vs, isVectorSelector := queryHints.CurrentNode.(*parser.VectorSelector)
```

Comment:
Shouldn't we handle `MatrixSelector` as well?

Reply:
I don't think so (but I might be missing something). A MatrixSelector doesn't have an offset, its fields are `Range` and `VectorSelector`. The contained `VectorSelector` may have an offset, and will be handled by this logic.

Reply:
Hmm `MatrixSelector` does not embed `VectorSelector` but `Expr` which may contain `VectorSelector`. So I guess in case of `MatrixSelector` it would not successfully type assert.

Reply:
In other words in case of `MatrixSelector` we would need to type assert on it's `. VectorSelector` field

Reply:
I've restructured the code by folding the `canAttemptPushdown` into `tryPushDown`. This makes the conditions that we're asserting clearer. In particular, it's more visible that we only attempt pushdowns on `VectorSelector`, which handles your concern here.